File: librarymanagementsystem/src/main/java/com/wipro/librarymanagementsystem/service/UserService.java
package com.wipro.librarymanagementsystem.service;

import com.wipro.librarymanagementsystem.entity.User;
import com.wipro.librarymanagementsystem.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.Optional;

@Service
public class UserService {
    
    @Autowired
    private UserRepository userRepository;
    
    // Get all users
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }
    
    // Get user by ID
    public Optional<User> getUserById(Long id) {
        return userRepository.findById(id);
    }
    
    // Get user by username
    public Optional<User> getUserByUsername(String username) {
        return userRepository.findByUsername(username);
    }
    
    // Get user by email
    public Optional<User> getUserByEmail(String email) {
        return userRepository.findByEmail(email);
    }
    
    // Save user
    public User saveUser(User user) {
        // Check if username already exists
        if (userRepository.existsByUsername(user.getUsername())) {
            throw new RuntimeException("Username '" + user.getUsername() + "' already exists");
        }
        
        // Check if email already exists
        if (userRepository.existsByEmail(user.getEmail())) {
            throw new RuntimeException("Email '" + user.getEmail() + "' already exists");
        }
        
        return userRepository.save(user);
    }
    
    // Update user
    public User updateUser(Long id, User userDetails) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found with id: " + id));
        
        // Check if username is being changed and if new username exists
        if (!user.getUsername().equals(userDetails.getUsername()) && 
            userRepository.existsByUsername(userDetails.getUsername())) {
            throw new RuntimeException("Username '" + userDetails.getUsername() + "' already exists");
        }
        
        // Check if email is being changed and if new email exists
        if (!user.getEmail().equals(userDetails.getEmail()) && 
            userRepository.existsByEmail(userDetails.getEmail())) {
            throw new RuntimeException("Email '" + userDetails.getEmail() + "' already exists");
        }
        
        user.setUsername(userDetails.getUsername());
        user.setEmail(userDetails.getEmail());
        user.setFirstName(userDetails.getFirstName());
        user.setLastName(userDetails.getLastName());
        user.setPhone(userDetails.getPhone());
        user.setAddress(userDetails.getAddress());
        user.setMembershipType(userDetails.getMembershipType());
        user.setIsActive(userDetails.getIsActive());
        
        return userRepository.save(user);
    }
    
    // Delete user
    public void deleteUser(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found with id: " + id));
        userRepository.delete(user);
    }
    
    // Deactivate user (soft delete)
    public User deactivateUser(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found with id: " + id));
        user.setIsActive(false);
        return userRepository.save(user);
    }
    
    // Activate user
    public User activateUser(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found with id: " + id));
        user.setIsActive(true);
        return userRepository.save(user);
    }
    
    // Search users by name
    public List<User> searchUsersByName(String name) {
        return userRepository.searchByName(name);
    }
    
    // Get active users
    public List<User> getActiveUsers() {
        return userRepository.findByIsActiveTrue();
    }
    
    // Get users by membership type
    public List<User> getUsersByMembershipType(User.MembershipType membershipType) {
        return userRepository.findByMembershipType(membershipType);
    }
    
    // Get users with overdue books
    public List<User> getUsersWithOverdueBooks() {
        return userRepository.findUsersWithOverdueBooks();
    }
    
    // Get user statistics
    public long getActiveUsersCount() {
        return userRepository.countByIsActiveTrue();
    }
    
    public long getTotalUsersCount() {
        return userRepository.count();
    }
}

File: frontend-Librarymanagementsystem/src/app/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { User, UserStats } from '../models';
import { AuthService } from './auth.service';

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private readonly API_URL = 'http://localhost:8083/api/users';

  constructor(
    private http: HttpClient,
    private authService: AuthService
  ) { }

  // Get all users
  getAllUsers(): Observable<User[]> {
    const headers = this.authService.getAuthHeader();
    return this.http.get<User[]>(this.API_URL, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  // Get user by ID
  getUserById(id: number): Observable<User> {
    const headers = this.authService.getAuthHeader();
    return this.http.get<User>(`${this.API_URL}/${id}`, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  // Get user by username
  getUserByUsername(username: string): Observable<User> {
    const headers = this.authService.getAuthHeader();
    return this.http.get<User>(`${this.API_URL}/username/${username}`, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  // Search users by name
  searchUsers(name: string): Observable<User[]> {
    const headers = this.authService.getAuthHeader();
    const params = new HttpParams().set('name', name);
    return this.http.get<User[]>(`${this.API_URL}/search`, { params, headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  // Get active users only
  getActiveUsers(): Observable<User[]> {
    const headers = this.authService.getAuthHeader();
    return this.http.get<User[]>(`${this.API_URL}/active`, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  // Get users by membership type
  getUsersByMembershipType(type: string): Observable<User[]> {
    const headers = this.authService.getAuthHeader();
    return this.http.get<User[]>(`${this.API_URL}/membership/${type}`, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  // Get users with overdue books
  getUsersWithOverdueBooks(): Observable<User[]> {
    const headers = this.authService.getAuthHeader();
    return this.http.get<User[]>(`${this.API_URL}/overdue`, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  // Create new user
  createUser(user: User): Observable<User> {
    const headers = this.authService.getAuthHeader();
    return this.http.post<User>(this.API_URL, user, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  // Update user
  updateUser(id: number, user: User): Observable<User> {
    const headers = this.authService.getAuthHeader();
    return this.http.put<User>(`${this.API_URL}/${id}`, user, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  // Deactivate user
  deactivateUser(id: number): Observable<User> {
    const headers = this.authService.getAuthHeader();
    return this.http.put<User>(`${this.API_URL}/${id}/deactivate`, {}, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  // Activate user
  activateUser(id: number): Observable<User> {
    const headers = this.authService.getAuthHeader();
    return this.http.put<User>(`${this.API_URL}/${id}/activate`, {}, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  // Delete user
  deleteUser(id: number): Observable<void> {
    const headers = this.authService.getAuthHeader();
    return this.http.delete<void>(`${this.API_URL}/${id}`, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  // Get user statistics
  getUserStats(): Observable<UserStats> {
    const headers = this.authService.getAuthHeader();
    return this.http.get<UserStats>(`${this.API_URL}/stats`, { headers })
      .pipe(
        catchError(this.handleError)
      );
  }

  // Error handling
  private handleError = (error: any): Observable<never> => {
    console.error('UserService error:', error);
    
    // Check for authentication errors
    if (error.status === 401 || error.status === 403) {
      this.authService.handleAuthError();
    }
    
    throw error;
  }
}